### ISSUE TITLE
Render Textures

### ISSUE THREAD

**Issue by if0ne** [GiiGaTeam/GiiGaEngine#65]
- [x] Создание RenderTarget текстур
- [x] Создание DepthTarget текстур
- [x] Создание ShaderVisible текстур (Shader Resource View)
- [x] Хранение дескрипторов RTV, DSV, SRV, UAV
- [x] Загрузка данных в текстуру из массива/вектора байтов

**Comment by damibran** [comment]
Чем текстура должна отличаться от GPULocalBuffer?

В случае разделения View и Buffer, не хочется и здесь хранить View. Но где тогда? Если с ShaderVisible все понятно это ShaderResource, то вот с RenderTarget etc.?

**Comment by damibran** [comment]
Единственно, что нарушается это логика current state, поскольку он меняется не только при update

**Comment by if0ne** [comment]
> Чем текстура должна отличаться от GPULocalBuffer?

Формат обязателен, subresource-ы по типу мипов, и доп вьюшки может иметь RenderTarget/Depth

**Comment by if0ne** [comment]
> В случае разделения View и Buffer, не хочется и здесь хранить View. Но где тогда? Если с ShaderVisible все понятно это ShaderResource, то вот с RenderTarget etc.?

Я бы здесь хранил View, текстура может разнообразные вью вообще иметь. Можно сделать мапу ViewDesc - View

**Comment by damibran** [comment]
> Загрузка данных в текстуру из массива/вектора байтов

Тут так скорее всего не получится, потому что например DDS текстуры уже могут содержать mipmap (supports single layer textures, textures with mipmaps, cube maps, volume maps and texture arrays (in Direct3D 10/11)).

Нужно иметь промежуточный объект CPU данных текстуры, который будет возвращать AssetLoader (или кто-то другой).

Например [3dgep|ScratchImage](https://web.archive.org/web/20240920224802/https://www.3dgep.com/learning-directx-12-4/#Loading_Textures).

**Comment by if0ne** [comment]
> > Загрузка данных в текстуру из массива/вектора байтов
> 
> Тут так скорее всего не получится, потому что например DDS текстуры уже могут содержать mipmap (supports single layer textures, textures with mipmaps, cube maps, volume maps and texture arrays (in Direct3D 10/11)).
> 
> Нужно иметь промежуточный объект CPU данных текстуры, который будет возвращать AssetLoader (или кто-то другой).
> 
> Например [3dgep|ScratchImage](https://web.archive.org/web/20240920224802/https://www.3dgep.com/learning-directx-12-4/#Loading_Textures).

Не знаю зачем этот ScratchImage, в тул ките уже есть [лоадер DDS текстур,](https://github.com/microsoft/DirectXTK12/wiki/DDSTextureLoader) который может из памяти (указатель + размер) создать ресурс со всеми мипами, или куб мапу и т.д. Также там можно просто название файла дать, может так быстрее будет.

**Comment by damibran** [comment]
Так, а если не dds текстура? Делать преобразование всех текстур к dds?

**Comment by damibran** [comment]
Да, можно передавать поток байт и тип потока, если это DDS использовать DDSTextureLoader, если что-то другое то...

**Comment by damibran** [comment]
Да, можно передавать поток байт и тип потока, если это DDS использовать DDSTextureLoader, если что-то другое то...

Или можно сделать кучу перегрузок конструктора, на каждый тип.

**Comment by if0ne** [comment]
> Так, а если не dds текстура? Делать преобразование всех текстур к dds?

Смотря, что будет в классе Texture. Если там не будет много специфичных данных, то все лоадеры текстур должны выдавать ID3D12Resource и соответственно 1 конструктор нужен от ID3D12Resource. Для других типов можно использовать dx12utils есть UpdateSubresource метод, который на основе промежуточного GPU буфера может сделать текстуру также с мипами, и даже если это TextureArray. Промежуточный буфер можно замапить и memcpy сделать.

**Comment by damibran** [comment]
Для других типов нужно также использовать [WICTextureLoader](https://github.com/microsoft/DirectXTK12/wiki/WICTextureLoader) из TK. Он тоже умеет генерировать mip при загрузке

**Comment by damibran** [comment]
Я все таки думаю, нужно объединять GPULocalBuffer и Texture в один класс Resource.
Соответственно по информации из входящего D3D12_RESOURCE_DESC или его же в пришедшем Resource, будет меняться логика.
Главной задачей которого будет:
1. Объединение ресурса и всех View (опять же есть проблема с Index Vertex, просто map не получиться, но решаемо тот же Variant)
2. Отслеживание состояния
3. Реализация функции обновления содержимого
